Doc/howto/a-conceptual-overview-of-asyncio.rst

```diff
@@ -0,0 +1,501 @@
+added back to the event loops queue.
+
+This is a basic, yet reliable mental model.
+In practice, it's slightly more complex, but not by much.
+In part 2, we'll walk through the details that make this possible.
+
+**Unlike tasks, awaiting a coroutine does not hand control back to the event
+loop!**
+Wrapping a coroutine in a task first, then ``await``\ ing that would cede
+control.
+The behavior of ``await coroutine`` is effectively the same as invoking a
+regular, synchronous Python function.
+Consider this program::
+
+   import asyncio
+
+   async def coro_a():
+      print("I am coro_a(). Hi!")
+
+   async def coro_b():
+      print("I am coro_b(). I sure hope no one hogs the event loop...")
+
+   async def main():
+      task_b = asyncio.create_task(coro_b())
+      num_repeats = 3
+      for _ in range(num_repeats):
+         await coro_a()
+      await task_b
+
+   asyncio.run(main())
+
+The first statement in the coroutine ``main()`` creates ``task_b`` and places
+it on the event loop's queue.
+Then, ``coro_a()`` is repeatedly ``await``\ ed. Control never cedes to the
+event loop which is why we see the output of all three ``coro_a()``
+invocations before ``coro_b()``'s output:
+
+.. code-block:: none
+
+   I am coro_a(). Hi!
+   I am coro_a(). Hi!
+   I am coro_a(). Hi!
+   I am coro_b(). I sure hope no one hogs the event loop...
+
+If we change ``await coro_a()`` to ``await asyncio.create_task(coro_a())``, the
+behavior changes.
+The coroutine ``main()`` cedes control to the event loop with that statement.
+The event loop then works through its queue, calling ``coro_b()`` and then
+``coro_a()`` before resuming the coroutine ``main()``.
+
+.. code-block:: none
+
+   I am coro_b(). I sure hope no one hogs the event loop...
+   I am coro_a(). Hi!
+   I am coro_a(). Hi!
+   I am coro_a(). Hi!
+
+------------------------------------------------
+A conceptual overview part 2: the nuts and bolts
+------------------------------------------------
+
+Part 2 goes into detail on the mechanisms :mod:`!asyncio` uses to manage
+control flow.
+This is where the magic happens.
+You'll come away from this section knowing what ``await`` does behind the scenes
+and how to make your own asynchronous operators.
+
+================================================
+coroutine.send(), await, yield and StopIteration
+================================================
+
+:mod:`!asyncio` leverages 4 components to pass around control.
+
+:meth:`coroutine.send(arg) <generator.send>` is the method used to start or
+resume a coroutine.
+If the coroutine was paused and is now being resumed, the argument ``arg``
+will be sent in as the return value of the ``yield`` statement which originally
+paused it.
+If the coroutine is being used for the first time (as opposed to being resumed)
+``arg`` must be ``None``.
+
+:ref:`yield <yieldexpr>`, like usual, pauses execution and returns control
+to the caller.
+In the example below, the ``yield``, on line 3, is called by
```

Comment:
Let's go ahead and present the example code before we discuss it.

Reply:
Ah, ya. Good call! Done 👍